contracts/zero-ex/contracts/src/features/ERC721OrdersFeature.sol

```diff
@@ -0,0 +1,1112 @@
+            order.direction == LibERC721Order.TradeDirection.SELL_721,
+            "ERC721OrdersFeature::_validateSellOrder/WRONG_TRADE_DIRECTION"
+        );
+        // The taker must either be unspecified in the order, or it must 
+        // be equal to `msg.sender`.
+        rrequire(
+            order.taker == address(0) || order.taker == msg.sender,
+            LibERC721OrdersRichErrors.OnlyTakerError(msg.sender, order.taker)
+        );
+        // Check that the order is valid and has not expired, been cancelled,
+        // or been filled.
+        LibERC721Order.OrderStatus status = getERC721OrderStatus(order);
+        rrequire(
+            getERC721OrderStatus(order) == LibERC721Order.OrderStatus.FILLABLE,
+            LibERC721OrdersRichErrors.OrderNotFillableError(
+                order.maker,
+                order.nonce,
+                uint8(status)
+            )
+        );
+        // Check the signature.
+        validateERC721OrderSignature(order, signature);
+    }
+
+    function _validateBuyOrder(
+        LibERC721Order.ERC721Order memory order,
+        LibSignature.Signature memory signature,
+        address taker,
+        uint256 erc721TokenId
+    )
+        private
+        view
+    {
+        // Order must be buying the ERC721 asset.
+        require(
+            order.direction == LibERC721Order.TradeDirection.BUY_721,
+            "ERC721OrdersFeature::_validateBuyOrder/WRONG_TRADE_DIRECTION"
+        );
+        // The taker must either be unspecified in the order, or it must 
+        // be equal to `msg.sender`.
+        rrequire(
+            order.taker == address(0) || order.taker == taker,
+            LibERC721OrdersRichErrors.OnlyTakerError(taker, order.taker)
+        );
+        // Check that the order is valid and has not expired, been cancelled,
+        // or been filled.
+        LibERC721Order.OrderStatus status = getERC721OrderStatus(order);
+        rrequire(
+            getERC721OrderStatus(order) == LibERC721Order.OrderStatus.FILLABLE,
+            LibERC721OrdersRichErrors.OrderNotFillableError(
+                order.maker,
+                order.nonce,
+                uint8(status)
+            )
+        );
+        // Check that the asset with the given token ID satisfies the properties
+        // specified by the order.
+        validateERC721OrderProperties(order, erc721TokenId);
+        // Check the signature.
+        validateERC721OrderSignature(order, signature);
+    }
+
+    function _payFees(
+        LibERC721Order.ERC721Order memory order,
+        address payer,
+        bool useNativeToken
+    )
+        private
+        returns (uint256 totalFeesPaid)
+    {
+        for (uint256 i = 0; i < order.fees.length; i++) {
+            LibERC721Order.Fee memory fee = order.fees[i];
+            
+            require(
+                fee.recipient != address(this), 
+                "ERC721OrdersFeature::_payFees/RECIPIENT_CANNOT_BE_EXCHANGE_PROXY"
```

Comment:
Probably not, but I have this here for the same reason we have this check in `FixinTokenSpender` and such, i.e. we don't want to avoid rogue self-calls as much as possible